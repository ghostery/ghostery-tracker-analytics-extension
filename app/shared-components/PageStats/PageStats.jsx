import React from 'react';
import PropTypes from 'prop-types';
import ClassNames from 'classnames';
import Nav from 'react-bootstrap/Nav';
import OverlayTrigger from 'react-bootstrap/OverlayTrigger';
import Tooltip from 'react-bootstrap/Tooltip';

import { convertBytes, truncateString } from '../../../src/utils/convert';
import './PageStats.scss';

let tooltipRefPointer;
let messageCreatorsPointer;

const dismissTabIconExplainer = () => (
  messageCreatorsPointer.markTabIconExplainerAcknowledged()
);

const dismissTabIconExplainerAndOpenApp = () => {
  dismissTabIconExplainer();
  messageCreatorsPointer.openApp();
};

const handleTrackerParentageIconClick = () => {
  messageCreatorsPointer.openApp(3);
  messageCreatorsPointer.sendMetrics({ type: 'tracker_parentage_visit' });
};

const minimizeButtonWithTooltip = () => (
  <OverlayTrigger
    container={tooltipRefPointer}
    placement="bottom"
    delay={500}
    overlay={(
      <Tooltip className="tooltipLeft">Minimize panel</Tooltip>
    )}
  >
    <img
      src={[chrome.extension.getURL('dist/images/shared/minimize-icon.svg')]}
      className="PageStats__panelButton infoIcon"
      width="14"
      height="14"
      alt="Minimize panel"
      onClick={messageCreatorsPointer.togglePanel}
    />
  </OverlayTrigger>
);

const tabIcon = (explainerActive = false) => (
  <img
    src={[chrome.extension.getURL('dist/images/shared/open-app-icon.svg')]}
    className="PageStats__panelButton infoIcon"
    width="14"
    height="14"
    alt="Open Insights tab"
    onClick={explainerActive
      ? dismissTabIconExplainerAndOpenApp
      : () => messageCreatorsPointer.openApp()}
  />
);

const tabIconWithTooltip = () => (
  <OverlayTrigger
    container={tooltipRefPointer}
    placement="bottom"
    delay={500}
    overlay={(
      <Tooltip className="tooltipLeft">Open Insights tab</Tooltip>
    )}
  >
    {tabIcon()}
  </OverlayTrigger>
);

function tabIconWithExplainer() {
  const descriptionText = 'Click here to view the Timeline and expanded dashboard.';
  const buttonText = 'Got it';

  return (
    <div className="PageStats__tabIconExplainerContainer">
      {tabIcon(true)}
      <div className="PageStats__tabIconExplainer">
        <div className="PageStats__iconAndDescriptionContainer">
          <img
            src={[chrome.extension.getURL('dist/images/shared/open-app-icon.svg')]}
            width="45"
            height="45"
            alt=""
          />
          <div className="description">
            {descriptionText}
          </div>
        </div>
        <div className="PageStats__ctaButton" onClick={dismissTabIconExplainer}>
          <span className="text">{buttonText}</span>
        </div>
      </div>
    </div>
  );
}

const trackerParentageIcon = () => (
  <div className="PageStats__trackerParentageButton" onClick={handleTrackerParentageIconClick}>
    <img
      src={[chrome.extension.getURL('dist/images/shared/tracker-parentage-icon.svg')]}
      width="19"
      height="20"
      alt="View Tracker Parentage"
    />
  </div>
);

const trackerParentageIconWithTooltip = () => (
  <OverlayTrigger
    container={tooltipRefPointer}
    placement="bottom"
    delay={500}
    overlay={(
      <Tooltip className="tooltipLeft">View Tracker Parentage</Tooltip>
    )}
  >
    {trackerParentageIcon()}
  </OverlayTrigger>
);

const PageStats = ({
  parentTabUrl,
  handleClick,
  panel,
  tabIconExplainerAcknowledged,
  trackerCount,
  requestsBySize,
  tooltipRef,
  messageCreators,
}) => {
  // export arguments needed by helper functions to their scope (i.e. file scope)
  // this keeps the helper functions calls clean and litter-free
  tooltipRefPointer = tooltipRef;
  messageCreatorsPointer = messageCreators;

  const numPageRequests = requestsBySize._tracker ? requestsBySize._tracker.count : 0;
  const totalTrackerSize = requestsBySize._tracker ? requestsBySize._tracker.size : 0;
  const convertedTrackerSize = convertBytes(totalTrackerSize);

  const pageStatsContainer = ClassNames('PageStats__container', { panel });
  const pageStats = ClassNames(
    'PageStats',
    'd-flex',
    'align-items-center',
    { panel, 'justify-content-between': panel, 'justify-content-around': !panel },
  );

  return (
    <div className={pageStatsContainer}>
      <div className={pageStats}>
        {parentTabUrl !== '' && (
          <Nav.Link onClick={handleClick}>
            {truncateString(parentTabUrl)}
          </Nav.Link>
        )}
        <div>{`Trackers: ${trackerCount}`}</div>
        <div>{`Tracker Requests: ${numPageRequests}`}</div>
        <div>{`Total Tracker Size: ${convertedTrackerSize}`}</div>
        {panel && (
          <div className="d-flex align-items-center justify-content-between">
            {minimizeButtonWithTooltip()}
            {
              (tabIconExplainerAcknowledged && tabIconWithTooltip())
              || tabIconWithExplainer()
            }
            {trackerParentageIconWithTooltip()}
          </div>
        )}
      </div>
    </div>
  );
};

PageStats.propTypes = {
  parentTabUrl: PropTypes.string,
  handleClick: PropTypes.func,
  panel: PropTypes.bool,
  tabIconExplainerAcknowledged: PropTypes.bool,
  requestsBySize: PropTypes.shape({
    _tracker: PropTypes.shape({
      size: PropTypes.number,
      count: PropTypes.number,
    }),
  }).isRequired,
  trackerCount: PropTypes.number.isRequired,
  tooltipRef: PropTypes.instanceOf(PropTypes.element),
  messageCreators: PropTypes.shape({
    markTabIconExplainerAcknowledged: PropTypes.func.isRequired,
    openApp: PropTypes.func.isRequired,
    sendMetrics: PropTypes.func.isRequired,
    togglePanel: PropTypes.func.isRequired,
  }).isRequired,
};

PageStats.defaultProps = {
  parentTabUrl: '',
  handleClick: null,
  panel: false,
  tabIconExplainerAcknowledged: false,
  tooltipRef: null,
};

export default PageStats;
