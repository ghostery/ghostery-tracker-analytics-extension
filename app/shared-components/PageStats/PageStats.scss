@import '../../utils/scss/color-variables.scss';

.PageStats__container {
  background-image: $foreground-gradient;
  border-radius: 5px;
  padding-bottom: 1px;
  height: 24px;

  &.panel {
    padding-bottom: 0;
    width: 500px;
  }
}

.PageStats {
  height: 24px;
  font-weight: 600;

  &.panel {
    padding: 0 10px 0 30px;
  }

  a {
    padding: 0;
    text-decoration: none;
    color: $white;
  }
}

.PageStats__trackerParentageButton {
  margin: 0 -11px 0 6px;
  box-shadow: inset 3px 0 4px -1px $opaque-background-half;
  background-color: $white;
  border-radius: 0 5px 5px 0;
  width: 35px;
  height: 24px;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;
}

.PageStats__panelButton {
  margin: 0 5px;
  border: 1px solid transparent;
  border-radius: 8px;
  cursor: pointer;

  &:hover {
    border: 1px solid $ghosty-blue;
  }
}

.PageStats__tabIconExplainerContainer {
  position: relative;
  top: 2px;
}

.PageStats__tabIconExplainer {
  position: absolute;
  z-index: 1;
  top: 34px;
  left: -160px;
  opacity: .98;
  box-shadow: 0 2px 4px 0 $opaque-background-half;
  background-image: $background-button-gradient;
  border: solid 1px $ghosty-blue;
  border-radius: 6px;
  width: 345px;
  height: 145px;
}

// Implements the triangle arrow that points to the tab icon

.PageStats__tabIconExplainer::after,
.PageStats__tabIconExplainer::before {
  position: absolute;
  bottom: 100%;
  left: 50%;
  border: solid;
  content: ' ';
}

.PageStats__tabIconExplainer::after {
  margin-left: -8px;
  border-color: $transparent $transparent $halfway-along-background-gradient;
  border-width: 0 8px 14px;
}

.PageStats__tabIconExplainer::before {
  margin-left: -10px;
  border-color: $transparent $transparent $ghosty-blue;
  border-width: 0 10px 16px;
}
// End of the triangle-arrow-related stuff

.PageStats__iconAndDescriptionContainer {
  margin-top: 20px;
  display: flex;
  align-items: center;
  justify-content: space-evenly;

  .description {
    width: 221px;
    height: 60px;
    font-size: 14px;
    line-height: 2.14;
  }
}

.PageStats__ctaButton {
  margin-top: 10px;
  margin-left: 98px;
  background-image: linear-gradient(110deg, $very-light-grey 1%, $white 100%);
  border-radius: 4px;
  width: 105px;
  height: 38px;
  display: flex;
  align-items: center;
  justify-content: center;
  cursor: pointer;

  &:hover {
    background-image: linear-gradient(110deg, $medium-grey 1%, $light-grey 100%);
  }

  .text {
    font-size: 14px;
    color: $topaz-blue;
  }
}
