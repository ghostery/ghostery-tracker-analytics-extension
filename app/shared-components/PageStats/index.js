import { connect } from 'react-redux';

import PageStats from './PageStats.jsx';
import { markTabIconExplainerAcknowledged, openApp, sendMetrics, togglePanel } from '../../store/messages/messageCreators';

const mapStateToProps = state => ({
  requestsBySize: state.PageInfoReducer.requestsBySize,
  trackerCount: state.PageInfoReducer.trackerCount,
  tooltipRef: state.UserInterfaceReducer.tooltipRef,
  tabIconExplainerAcknowledged: state.SettingsReducer.localSettings.tabIconExplainerAcknowledged,
  messageCreators: { markTabIconExplainerAcknowledged, openApp, sendMetrics, togglePanel },
});

export default connect(mapStateToProps)(PageStats);
