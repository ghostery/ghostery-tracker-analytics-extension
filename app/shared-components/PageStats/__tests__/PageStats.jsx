import React from 'react';
import { shallow } from 'enzyme';
import chrome from 'sinon-chrome';
import path from 'path';
import fakeProps from 'react-fake-props';

import PageStats from '../PageStats';

describe('app/shared-components/PageStats/', () => {
  describe('Snapshot tests with shallow', () => {
    test('PageStats component should render', () => {
      global.chrome = chrome;
      const componentPath = path.join(__dirname, '../PageStats.jsx');
      const props = fakeProps(componentPath);

      const component = shallow(<PageStats {...props} />);
      expect(component).toMatchSnapshot();
    });
  });
});
