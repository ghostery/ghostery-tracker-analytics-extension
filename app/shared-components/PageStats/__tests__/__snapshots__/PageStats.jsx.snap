// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/shared-components/PageStats/ Snapshot tests with shallow PageStats component should render 1`] = `
<div
  className="PageStats__container"
>
  <div
    className="PageStats d-flex align-items-center justify-content-around"
  >
    <div>
      Trackers: 1
    </div>
    <div>
      Tracker Requests: 0
    </div>
    <div>
      Total Tracker Size: 0 B
    </div>
  </div>
</div>
`;
